tsdb/db.go

```diff
@@ -1476,32 +1476,42 @@ func (db *DB) CleanTombstones() (err error) {
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
-	newUIDs := []ulid.ULID{}
-	defer func() {
-		// If any error is caused, we need to delete all the new directory created.
-		if err != nil {
-			for _, uid := range newUIDs {
-				dir := filepath.Join(db.Dir(), uid.String())
-				if err := os.RemoveAll(dir); err != nil {
-					level.Error(db.logger).Log("msg", "failed to delete block after failed `CleanTombstones`", "dir", dir, "err", err)
-				}
-			}
-		}
-	}()
+	var (
+		cleanUpCompleted = false
+		blocks           []*Block
+	)
 
-	db.mtx.RLock()
-	blocks := db.blocks[:]
-	db.mtx.RUnlock()
+	// Repeat cleanup until there is no tombstones left
+	for !cleanUpCompleted {
+		cleanUpCompleted = true
 
-	for _, b := range blocks {
-		if uid, er := b.CleanTombstones(db.Dir(), db.compactor); er != nil {
-			err = errors.Wrapf(er, "clean tombstones: %s", b.Dir())
-			return err
-		} else if uid != nil { // New block was created.
-			newUIDs = append(newUIDs, *uid)
+		blocks = db.Blocks()
+		for _, b := range blocks {
```

Comment:
blocks will be invalidated after the reload, so you should check that the b is still in the current list.

Reply:
Sorry, I'm not sure I understood your comment. The `break` at the end of the inner loop will make sure that I'm not looping over the same blocks after a reload.